### PR TITLE
Storage UI for user documents

### PR DESCRIPTION
**Pull request by guperrot**
Please have a look at our [guidelines for contributions](https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/CONTRIBUTING.md) and consider the following before you submit the PR:

* [ ] Has `CHANGELOG.md` been updated?
* [x] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?
* [x] Did you check UI tests on the sample app? They are not executed on CI.

## Description

Storage UI for user documents

image
image

### CODE REVIEW

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 1]
```diff
@@ -54,17 +61,38 @@ public View getView(final int position, View convertView, ViewGroup parent) {
...
                     mList.remove(position);
                     notifyDataSetChanged();
                 }
             });
         }
         return view;
     }
+
+    public ArrayList<String> loadArray(Context context, String name) {
+        SharedPreferences sp = mContext.getSharedPreferences(name, Context.MODE_PRIVATE);
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]:
```suggestion
        SharedPreferences sharedPreferences = mContext.getSharedPreferences(name, Context.MODE_PRIVATE);
```

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 2]
```diff
@@ -54,17 +61,38 @@ public View getView(final int position, View convertView, ViewGroup parent) {
...
+        ArrayList<String> list = new ArrayList<>();
+        int size = sp.getInt("Status_size", 0);
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]:
```suggestion
        int size = sharedPreferences.getInt("Status_size", 0);
```

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 3]
```diff
@@ -54,17 +61,38 @@ public View getView(final int position, View convertView, ViewGroup parent) {
...
+        for (int i = 0; i < size; i++) {
+            list.add(sp.getString("Status_" + i, null));
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]:
```suggestion
            list.add(sharedPreferences.getString("Status_" + i, null));
```

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 4]
```diff
@@ -54,17 +61,38 @@ public View getView(final int position, View convertView, ViewGroup parent) {
...
+                    Storage.delete(Constants.USER, StorageActivity.mUserDocumentList.get(position));
                     mList.remove(position);
                     notifyDataSetChanged();
                 }
             });
         }
         return view;
     }
+
+    public ArrayList<String> loadArray(Context context, String name) {
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]: I would rename signature to loadArrayFromPreferences(context, arrayKey)

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/NewUserDocumentActivity.java [hunk 5]
```diff
@@ -11,21 +11,32 @@
...
     private final List<TypedPropertyFragment> mProperties = new ArrayList<>();
 
+    private EditText mEditDocumentId;
+
     @Override
     protected void onCreate(@Nullable Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_new_user_document);
+
+        mEditDocumentId = findViewById(R.id.user_document_id);
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/NewUserDocumentActivity.java]: Please remove empty lines

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 6]
```diff
@@ -42,16 +47,17 @@
 
 public class StorageActivity extends AppCompatActivity {
 
-    private ListView mListView;
+    public static ListView mListView;
+
+    public static ArrayList<String> mUserDocumentList = new ArrayList<String>();
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: public variables do not start with "m": https://source.android.com/setup/contribute/code-style#follow-field-naming-conventions

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 7]
```diff
@@ -42,16 +47,17 @@
...
+
+    public static CustomItemAdapter adapterUser;
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: static variables start with "s"

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 8]
```diff
@@ -75,6 +81,24 @@ public void accept(PaginatedDocuments<TestDocument> documents) {
...
+        mUserDocumentContents.clear();
+        if (adapterUser == null) {
+            String accountId = MainActivity.sSharedPreferences.getString(ACCOUNT_ID, null);
+            if (accountId != null) {
+                Storage.list(Constants.USER, Map.class).thenAccept(new AppCenterConsumer<PaginatedDocuments<Map>>() {
+                    @Override
+                    public void accept(PaginatedDocuments<Map> documents) {
+                        for (Document<Map> document : documents.getCurrentPage().getItems()) {
+                            mUserDocumentList.add(document.getId());
+                            mUserDocumentContents.add(Utils.getGson().toJson(document.getDocument()));
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Can getDocument return null?

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 9]
```diff
@@ -173,6 +199,42 @@ private void updateStorageType(int position) {
         }
     }
 
+    public boolean saveArray(ArrayList<String> list, String name) {
+        SharedPreferences sp = getSharedPreferences(name, Context.MODE_PRIVATE);
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]:
```suggestion
        SharedPreferences sharedPreferences = getSharedPreferences(name, Context.MODE_PRIVATE);
```

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 10]
```diff
@@ -173,6 +199,42 @@ private void updateStorageType(int position) {
...
+        SharedPreferences.Editor editor = sp.edit();
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]:
```suggestion
        SharedPreferences.Editor editor = sharedPreferences.edit();
```

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 11]
```diff
@@ -173,6 +199,42 @@ private void updateStorageType(int position) {
         }
     }
 
+    public boolean saveArray(ArrayList<String> list, String name) {
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: I would rename it to saveArrayToPreferences(array, arrayKey)

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 12]
```diff
@@ -173,6 +199,42 @@ private void updateStorageType(int position) {
...
+            mUserDocumentList.clear();
+            mUserDocumentContents.clear();
+            String accountId = MainActivity.sSharedPreferences.getString(ACCOUNT_ID, null);
+            if (accountId != null) {
+                Storage.list(Constants.USER, Map.class).thenAccept(new AppCenterConsumer<PaginatedDocuments<Map>>() {
+                    @Override
+                    public void accept(PaginatedDocuments<Map> documents) {
+                        for (Document<Map> document : documents.getCurrentPage().getItems()) {
+                            mUserDocumentList.add(document.getId());
+                            mUserDocumentContents.add(Utils.getGson().toJson(document.getDocument()));
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Can getDocument return null?

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/UserDocumentDetailActivity.java [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+import static com.microsoft.appcenter.sasquatch.SasquatchConstants.USER_DOCUMENT_CONTENTS;
+import static com.microsoft.appcenter.sasquatch.SasquatchConstants.USER_DOCUMENT_LIST;
+
+public class UserDocumentDetailActivity extends AppCompatActivity {
+
+    @Override
+    protected void onCreate(@Nullable Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_user_document_detail);
+
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/UserDocumentDetailActivity.java]: Empty line

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+            android:layout_marginStart="0sp"
+            android:layout_marginLeft="0sp"
+            android:layout_marginTop="35sp"
+            android:layout_weight="1"
+            android:textSize="20sp" />
+
+    </LinearLayout>
+
+
+</LinearLayout>
```
Comment (annakocheshkova) [inline apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: Please run formatter on this file

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java (hunk 12 above)
Comment (guperrot) [reply apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Yes on error, good point.
We need to do error management and maybe show a toast.

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml (hunk 14 above)
Comment (guperrot) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: And remove blank lines.

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 15]
```diff
@@ -42,16 +47,17 @@
 
 public class StorageActivity extends AppCompatActivity {
 
+    public static ArrayList<String> sUserDocumentList = new ArrayList<String>();
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: why do you need a static state? you clear it in onCreate, its useless

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/NewUserDocumentActivity.java [hunk 16]
```diff
@@ -51,4 +60,16 @@ private void addProperty() {
...
     }
+
+    public void save(View view) {
+        Map<String, Object> document = new LinkedHashMap<>();
+        for (TypedPropertyFragment property : mProperties) {
+            property.setGenericProperty(document);
+        }
+
+        /* TODO use .thenAccept and Toast message whether success or error. */
+        /* TODO replace "id" by the one from the text edit. */
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/NewUserDocumentActivity.java]: The second todo is solved so no need to keep the second comment.

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 17]
```diff
@@ -75,6 +81,24 @@ public void accept(PaginatedDocuments<TestDocument> documents) {
...
         });
 
+        /* List the user documents. */
+        sUserDocumentList.clear();
+        mUserDocumentContents.clear();
+        if (mAdapterUser == null) {
+            String accountId = MainActivity.sSharedPreferences.getString(ACCOUNT_ID, null);
+            if (accountId != null) {
+                Storage.list(Constants.USER, Map.class).thenAccept(new AppCenterConsumer<PaginatedDocuments<Map>>() {
+                    @Override
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]:
```suggestion
                    
                    @Override
```
blank lines before annotations

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 18]
```diff
@@ -173,6 +199,42 @@ private void updateStorageType(int position) {
         }
     }
 
+    public boolean saveArrayToPreferences(ArrayList<String> list, String name) {
+        MainActivity.sSharedPreferences = getSharedPreferences(name, Context.MODE_PRIVATE);
+        SharedPreferences.Editor editor = MainActivity.sSharedPreferences.edit();
+        editor.putInt("Status_size", list.size());
+
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: no block blank line without a comment

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml [hunk 19]
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="utf-8"?><!--
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: Why do you need a new layout? It does not seem to be used.

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+  ~ Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: where is this file used?

Comment (zhangcc01) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: @guperrot this file is used for user document detail in UserDocumentDetailActivity.java

Comment (guperrot) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: Is there any difference with app document view or is it copy paste? If it's the same we don't 2 code bases.

File: versions.gradle [hunk 21]
```diff
@@ -8,7 +8,7 @@
 ext {
     versionCode = 45
     versionName = '1.11.5'
-    minSdkVersion = 16
+    minSdkVersion = 19
```
Comment (MatkovIvan) [inline versions.gradle]: Please revert

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 22]
```diff
@@ -255,6 +259,27 @@ public void accept(PaginatedDocuments<Map> documents) {
...
+            editor.remove("Status_" + i);
+            editor.putString("Status_" + i, list.get(i));
+        }
+        editor.commit();
+    }
+
+    private ArrayList<String> loadArrayFromPreferences(String name) {
+        MainActivity.sSharedPreferences = getSharedPreferences(name, Context.MODE_PRIVATE);
+        ArrayList<String> list = new ArrayList<>();
+        int size = MainActivity.sSharedPreferences.getInt("Status_size", 0);
```
Comment (MatkovIvan) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Use const

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java [hunk 23]
```diff
@@ -0,0 +1,86 @@
...
+import com.microsoft.appcenter.sasquatch.R;
+import com.microsoft.appcenter.storage.Utils;
+import com.microsoft.appcenter.storage.models.Document;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AppDocumentListAdapter extends RecyclerView.Adapter<AppDocumentListAdapter.AppDocumentListHolder> {
+
+    private Context mContext;
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java]: blank lines between fields

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+    private List<Document<TestDocument>> mList;
+    private OnItemClickListener listener;
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java]: mListener

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java [hunk 25]
```diff
@@ -0,0 +1,86 @@
...
+
+    public AppDocumentListAdapter(Context context, List<Document<TestDocument>> list) {
+        this.mContext = context;
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java]: don't use this

File: apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java [hunk 26]
```diff
@@ -0,0 +1,86 @@
...
+        return mList.get(position).getId();
+    }
+
+    @Override
+    public int getItemCount() {
+        return mList.size();
+    }
+
+    public void setOnItemClickListener(OnItemClickListener listener) {
+        this.listener = listener;
```
Comment (guperrot) [inline apps/sasquatch/src/main/java/com/microsoft/appcenter/sasquatch/activities/AppDocumentListAdapter.java]: mListener = listener

File: apps/sasquatch/src/main/res/values/strings.xml [hunk 27]
```diff
@@ -99,4 +99,8 @@
         <item>Invalid</item>
     </string-array>
     <string name="number_of_logs" tools:ignore="MissingTranslation">Number of logs: %d</string>
+    <string name="message_whether_success" tools:ignore="MissingTranslation">Message whether success.</string>
```
Comment (guperrot) [inline apps/sasquatch/src/main/res/values/strings.xml]: Please be more specific about those messages.

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 28]
```diff
@@ -5,66 +5,94 @@
...
 
 import java.util.ArrayList;
 
-public class CustomItemAdapter extends BaseAdapter implements ListAdapter {
+public class CustomItemAdapter extends RecyclerView.Adapter<CustomItemAdapter.CustomItemAdapterHolder> {
 
     private ArrayList<String> mList;
-
     private Context mContext;
+    private CustomItemAdapter.OnItemClickListener listener;
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]: blankline before and rename to mListener and don't use `this`

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java [hunk 29]
```diff
@@ -5,66 +5,94 @@
...
+        CustomItemAdapterHolder(@NonNull View itemView) {
+            super(itemView);
+            listItemText = itemView.findViewById(R.id.property);
+            deleteBtn = itemView.findViewById(R.id.delete_button);
         }
-        return view;
+    }
+
+    public interface OnItemClickListener {
+        void onItemClick(int position);
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/CustomItemAdapter.java]: blankline before method

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 30]
```diff
@@ -42,38 +52,138 @@
...
+
+    private void updateUserDocuments(List<Document<Map>> documents) {
+        if (documents == null)
+            return;
+
+        for (Document<Map> document : documents) {
+            sUserDocumentList.add(String.format("%s_%s", document.isFromCache() ? CACHED_PREFIX : REMOTE_PREFIX, document.getId()));
+            mUserDocumentContents.add(Utils.getGson().toJson(document.getDocument()));
+        }
+        saveArrayToPreferences(sUserDocumentList, USER_DOCUMENT_LIST);
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Don't cache lists

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 31]
```diff
@@ -146,33 +283,60 @@ public void onClick(DialogInterface dialog, int which) {
...
+            editor.remove("Status_" + i);
+            editor.putString("Status_" + i, list.get(i));
+        }
+        editor.commit();
+    }
+
+    private ArrayList<String> loadArrayFromPreferences(String name) {
+        MainActivity.sSharedPreferences = getSharedPreferences(name, Context.MODE_PRIVATE);
+        ArrayList<String> list = new ArrayList<>();
+        int size = MainActivity.sSharedPreferences.getInt("Status_size", 0);
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: Don't store lists to cache.

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml (hunk 20 above)
Comment (guperrot) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: ping

File: versions.gradle (hunk 21 above)
Comment (guperrot) [reply versions.gradle]: Yes please revert indeed.

Comment (guperrot) [reply versions.gradle]: And catch up with develop.

**Comment by guperrot** [comment]
Please merge with develop and fix the app after merging. The CI fails and always merge locally before running. Please pay attention to the red build details.

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml (hunk 20 above)
Comment (annakocheshkova) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: @guperrot yes we already saw that there are two similar activities, will merge them into one

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java (hunk 30 above)
Comment (annakocheshkova) [reply apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: :+1: already discussed this with storage team.

File: apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml (hunk 20 above)
Comment (annakocheshkova) [reply apps/sasquatch/src/projectDependency/res/layout/activity_user_document_detail.xml]: In a new PR: https://github.com/Microsoft/AppCenter-SDK-Android/pull/1078

File: apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java [hunk 32]
```diff
@@ -48,41 +48,41 @@
 
     private RecyclerView mListView;
 
-    private Boolean isLoading = false;
+    private Boolean mIsLoading = false;
```
Comment (guperrot) [inline apps/sasquatch/src/projectDependency/java/com/microsoft/appcenter/sasquatch/activities/StorageActivity.java]: do we use null here? if not rather use boolean

**Pull request merged**